### PR TITLE
fix: combined publish flow for CDK/SDM

### PR DESCRIPTION
**Pull request by ChristoGrab**
## What

- Consolidates the steps to publish the CDK to PyPI and Source Declarative Manifest to Dockerhub into a single workflow.
- Updates the path of the uploaded build artifacts to match the path used during the build
- Updates the SDM Dockerfile to use the build artifacts

> [!WARNING]
99% sure the workflow_dispatch trigger is still broken. I ran into issues with the input version not being respected by the build step, leading to all sorts of shenanigans. Decided to prioritize getting the release trigger working to unblock us on publishing new CDK/SDM versions. Fow now, the following should run and work:

  - Build: any push
  - Publish (PYPI + DOCKERHUB): new releases

## Summary by CodeRabbit

- **New Features**
  - Enhanced packaging and publishing workflow for Python projects, including new version input and improved artifact management.
  - Added a new job for publishing Docker images to DockerHub, ensuring multi-platform support and tag management.

- **Improvements**
  - Streamlined Dockerfile for Python projects, simplifying installation by using pre-built wheel files.

- **Bug Fixes**
  - Removed unnecessary build steps in the Dockerfile to optimize the build process.

### CODE REVIEW

File: .github/workflows/pypi_publish.yml [hunk 1]
```diff
@@ -40,8 +53,76 @@ jobs:
...
+        run: |
+          echo "Version set to ${{ github.ref_name }}"
+          echo "VERSION=${{ github.ref_name }}" >> $GITHUB_ENV
+
+      # We need to download the build artifact again because the previous job was on a different runner
+      - name: Download Build Artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: Packages-${{ github.run_id }}
+          path: dist
```
Comment (ChristoGrab) [inline .github/workflows/pypi_publish.yml:87]: Adding this step is necessary to ensure that the SDM image has access to the build artifact and can access the correct CDK version

File: Dockerfile [hunk 2]
```diff
@@ -13,6 +14,6 @@ RUN poetry config virtualenvs.create false \
 COPY airbyte_cdk ./airbyte_cdk
 
 # Build and install the package
-RUN poetry build && pip install dist/*.whl
+RUN pip install dist/*.whl
```
Comment (ChristoGrab) [inline Dockerfile:17]: Using poetry build here would overwrite the CDK version obtained from the build artifact

File: Dockerfile [hunk 3]
```diff
@@ -4,6 +4,7 @@ WORKDIR /airbyte/integration_code
 
 # Copy project files needed for build
 COPY pyproject.toml poetry.lock README.md ./
+COPY dist/*.whl ./dist/
```
Comment (ChristoGrab) [inline Dockerfile:7]: Copy the packages obtained from the build

File: .github/workflows/pypi_publish.yml [hunk 4]
```diff
@@ -40,8 +53,72 @@ jobs:
...
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Set Version (workflow_dispatch)
+        if: github.event_name == 'workflow_dispatch'
+        run: |
+          echo "Version set to ${{ github.event.inputs.version }}"
+          echo "VERSION=${{ github.event.inputs.version }}" >> $GITHUB_ENV
+
```
[the remark this answers is not in the document]
Comment (ChristoGrab) [reply .github/workflows/pypi_publish.yml:75]: @coderabbitai - Although that is generally a good practice, in this case the quotes would become part of the VERSION value, meaning the Docker tag would end up like:

airbyte/source-declarative-manifest:'6.7.0dev8'

File: .github/workflows/pypi_publish.yml [hunk 5]
```diff
@@ -1,37 +1,50 @@
...
           fetch-depth: 0
+          ref: ${{ github.event_name == 'workflow_dispatch' && format('v{0}', github.event.inputs.version) || github.ref }}
+
       - uses: hynek/build-and-inspect-python-package@v2
 
+      - uses: actions/upload-artifact@v4
+        with:
+          name: Packages-${{ github.run_id }}
+          path: |
+            /tmp/baipp/dist/*.whl
```
Comment (ChristoGrab) [inline .github/workflows/pypi_publish.yml:26]: Had to specify this path for uploading build artifacts as the /dist directory was not matching the path in the previous step

**Review by aaronsteers — APPROVED** [review]

File: .github/workflows/pypi_publish.yml [hunk 6]
```diff
@@ -40,8 +53,72 @@ jobs:
...
         with:
-          # Can be toggled at the repository level between `https://upload.pypi.org/legacy/` and `https://test.pypi.org/legacy/`
-          repository-url: ${{vars.PYPI_PUBLISH_URL}}
+          context: .
+          platforms: linux/amd64,linux/arm64
+          push: true
+          tags: |
+            airbyte/source-declarative-manifest:latest
+            airbyte/source-declarative-manifest:${{ env.VERSION }}
+            airbyte/source-declarative-manifest:${{ github.sha }}
```
Comment (natikgadzhi) [inline .github/workflows/pypi_publish.yml:124]: @aaronsteers and @ChristoGrab: do we really want to push builds for sha tags to public dockerhub repo? Wouldn't it be too noisy?

File: Dockerfile (hunk 3 above)
Comment (natikgadzhi) [reply Dockerfile:7]: A comment + link to code on what builds it and why we think it'll be there would help!

**Pull request merged**